polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
 
-  // We use two collections here as we need to preserve the order for insertion
-  // and this is easiest.
-  SmallPtrSet<Operation *, 8> willBeMovedSet;
-  SmallVector<Operation *, 8> opsToMove;
+  return true;
+}
 
+// Populate \p opsToMove with operations that can be hoisted out of the given
+// loop \p loop.
+template <typename T, typename = std::enable_if_t<llvm::is_one_of<
+                          T, scf::ParallelOp, AffineParallelOp>::value>>
+static void collectHoistableOperation(T loop,
+                                      SmallVectorImpl<Operation *> &opsToMove) {
   // Do not use walk here, as we do not want to go into nested regions and hoist
   // operations from there. These regions might have semantics unknown to this
   // rewriting. If the nested regions are loops, they will have been processed.
-  for (auto &block : loopBody) {
-    for (auto &op : block.without_terminator()) {
-      if (canBeParallelHoisted(&op, looplike, willBeMovedSet)) {
-        opsToMove.push_back(&op);
-        willBeMovedSet.insert(&op);
-      }
+  SmallPtrSet<Operation *, 8> willBeMoved;
+  for (Block &block : loop.getLoopBody()) {
+    for (Operation &op : block.without_terminator()) {
+      if (!canBeHoisted(op, loop, willBeMoved))
+        continue;
+      opsToMove.push_back(&op);
+      willBeMoved.insert(&op);
     }
   }
+}
 
-  // For all instructions that we found to be invariant, move outside of the
-  // loop.
-  if (opsToMove.size()) {
-    OpBuilder b(looplike);
-    Value cond = nullptr;
-    for (auto pair : llvm::zip(looplike.getLowerBound(),
-                               looplike.getUpperBound(), looplike.getStep())) {
-      auto val = b.create<arith::CmpIOp>(
-          looplike.getLoc(), CmpIPredicate::sle,
-          b.create<arith::AddIOp>(looplike.getLoc(), std::get<0>(pair),
-                                  std::get<2>(pair)),
-          std::get<1>(pair));
-      if (cond == nullptr)
-        cond = val;
-      else
-        cond = b.create<arith::AndIOp>(looplike.getLoc(), cond, val);
-    }
-    auto ifOp = b.create<scf::IfOp>(looplike.getLoc(), TypeRange(), cond);
-    looplike->moveBefore(ifOp.thenYield());
+void moveParallelLoopInvariantCode(scf::ParallelOp loop) {
+  SmallVector<Operation *, 8> opsToMove;
+  collectHoistableOperation(loop, opsToMove);
+  if (opsToMove.empty())
+    return;
+
+  // Move all operations we found to be invariant outside of the loop.
+  OpBuilder b(loop);
+  Value cond = nullptr;
+  for (auto pair :
+       llvm::zip(loop.getLowerBound(), loop.getUpperBound(), loop.getStep())) {
+    auto val = b.create<arith::CmpIOp>(
+        loop.getLoc(), arith::CmpIPredicate::sle,
+        b.create<arith::AddIOp>(loop.getLoc(), std::get<0>(pair),
+                                std::get<2>(pair)),
+        std::get<1>(pair));
+    cond =
+        cond ? (Value)b.create<arith::AndIOp>(loop.getLoc(), cond, val) : val;
```

Comment:
```suggestion
    const auto val = static_cast<Value>(b.create<arith::CmpIOp>(
        loop.getLoc(), arith::CmpIPredicate::sle,
        b.create<arith::AddIOp>(loop.getLoc(), std::get<0>(pair),
                                std::get<2>(pair)),
        std::get<1>(pair)));
    cond =
        cond ? stastic_cast<Value>(b.create<arith::AndIOp>(loop.getLoc(), cond, val)) : val;
```
Avoid C-like cast, make `val` `const` and make typing more clear

Reply:
Done, but I avoided the first static_cast by making the declared type on `val` a `const Value`

Reply:
I personally prefer the `static_cast` to avoid "hiding the cast", but LGTM